File: app/src/main/java/com/verbatoria/presentation/schedule/view/adapter/ScheduleAdapter.java
package com.verbatoria.presentation.schedule.view.adapter;

import android.content.Context;
import android.content.res.Resources;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.cleveroad.adaptivetablelayout.LinkedAdaptiveTableAdapter;
import com.cleveroad.adaptivetablelayout.ViewHolderImpl;
import com.remnev.verbatoriamini.R;
import com.verbatoria.business.schedule.datasource.IScheduleDataSource;
import com.verbatoria.business.schedule.models.ScheduleItemModel;
import com.verbatoria.presentation.schedule.view.adapter.viewholder.ColumnHeaderViewHolder;
import com.verbatoria.presentation.schedule.view.adapter.viewholder.RowHeaderViewHolder;
import com.verbatoria.presentation.schedule.view.adapter.viewholder.ScheduleItemViewHolder;
import com.verbatoria.presentation.schedule.view.adapter.viewholder.ScheduleTopLeftViewHolder;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * @author nikitaremnev
 */

public class ScheduleAdapter extends LinkedAdaptiveTableAdapter<ViewHolderImpl> {

    private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("dd.MM", Locale.ROOT);
    private static final SimpleDateFormat TIME_FORMAT = new SimpleDateFormat("HH:mm", Locale.ROOT);

    private static String[] ROWS_NAMES;

    private final LayoutInflater mLayoutInflater;

    private final int mHeaderColumnHeight;
    private final int mColumnWidth;
    private final int mRowHeight;
    private final int mHeaderRowWidth;

    private IScheduleDataSource<String, Date, Date, ScheduleItemModel> mDataSource;

    public ScheduleAdapter(Context context, IScheduleDataSource<String, Date, Date, ScheduleItemModel> dataSource) {
        mLayoutInflater = LayoutInflater.from(context);
        Resources resources = context.getResources();
        mHeaderColumnHeight = resources.getDimensionPixelSize(R.dimen.column_header_height);
        mRowHeight = resources.getDimensionPixelSize(R.dimen.row_height);
        mHeaderRowWidth = resources.getDimensionPixelSize(R.dimen.row_header_width);
        mColumnWidth = resources.getDimensionPixelSize(R.dimen.column_width);
        ROWS_NAMES = context.getResources().getStringArray(R.array.rows_names);
        mDataSource = dataSource;
    }

    @Override
    public int getRowCount() {
        return mDataSource.getRowsCount();
    }

    @Override
    public int getColumnCount() {
        return mDataSource.getColumnsCount();
    }

    @NonNull
    @Override
    public ViewHolderImpl onCreateItemViewHolder(@NonNull ViewGroup parent) {
        return new ScheduleItemViewHolder(mLayoutInflater.inflate(R.layout.item_schedule_cell, parent, false));
    }

    @NonNull
    @Override
    public ViewHolderImpl onCreateColumnHeaderViewHolder(@NonNull ViewGroup parent) {
        return new ColumnHeaderViewHolder(mLayoutInflater.inflate(R.layout.item_column_header, parent, false));
    }

    @NonNull
    @Override
    public ViewHolderImpl onCreateRowHeaderViewHolder(@NonNull ViewGroup parent) {
        return new RowHeaderViewHolder(mLayoutInflater.inflate(R.layout.item_row_header, parent, false));
    }

    @NonNull
    @Override
    public ViewHolderImpl onCreateLeftTopHeaderViewHolder(@NonNull ViewGroup parent) {
        return new ScheduleTopLeftViewHolder(mLayoutInflater.inflate(R.layout.item_schedule_top_left, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolderImpl viewHolder, int row, int column) {
        ScheduleItemModel scheduleItemModel = mDataSource.getItemData(row, column);
        ((ScheduleItemViewHolder) viewHolder).setChecked(scheduleItemModel.isSelected());
    }

    @Override
    public void onBindHeaderColumnViewHolder(@NonNull ViewHolderImpl viewHolder, int column) {
        ((ColumnHeaderViewHolder) viewHolder).setHeaderTitle(TIME_FORMAT.format(mDataSource.getColumnHeaderData(column)));
    }

    @Override
    public void onBindHeaderRowViewHolder(@NonNull ViewHolderImpl viewHolder, int row) {
        ((RowHeaderViewHolder) viewHolder).setHeaderTitle(
                String.format(ROWS_NAMES[row - 1], DATE_FORMAT.format(mDataSource.getRowHeaderData(row))));
    }

    @Override
    public void onBindLeftTopHeaderViewHolder(@NonNull ViewHolderImpl viewHolder) {

    }

    @Override
    public int getColumnWidth(int column) {
        return mColumnWidth;
    }

    @Override
    public int getHeaderColumnHeight() {
        return mHeaderColumnHeight;
    }

    @Override
    public int getRowHeight(int row) {
        return mRowHeight;
    }

    @Override
    public int getHeaderRowWidth() {
        return mHeaderRowWidth;
    }

    public void notifyItemChanged(int row, int column, boolean state) {
        mDataSource.getItemData(row, column).setSelected(state);
        notifyItemChanged(row, column);
    }
}

File: app/src/main/java/com/verbatoria/ui/schedule/ScheduleAdapter.kt
package com.verbatoria.ui.schedule

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import com.remnev.verbatoria.R
import com.verbatoria.domain.schedule.ScheduleDataSource
import com.verbatoria.infrastructure.extensions.formatToShortDate
import com.verbatoria.infrastructure.extensions.formatToTime
import com.verbatoria.ui.common.adaptivetablelayout.LinkedAdaptiveTableAdapter
import com.verbatoria.ui.common.adaptivetablelayout.ViewHolderImpl
import com.verbatoria.ui.schedule.item.ScheduleColumnHeaderViewHolder
import com.verbatoria.ui.schedule.item.ScheduleItemViewHolder
import com.verbatoria.ui.schedule.item.ScheduleRowHeaderViewHolder
import com.verbatoria.ui.schedule.item.ScheduleTopLeftViewHolder

/**
 * @author n.remnev
 */

private const val ROWS_COUNT = 7
private const val COLUMN_COUNT = 11

class ScheduleAdapter(
    context: Context,
    width: Int,
    height: Int,
    private val scheduleDataSource: ScheduleDataSource
): LinkedAdaptiveTableAdapter<ViewHolderImpl>() {

    private val headerColumnHeight: Int = context.resources.getDimensionPixelSize(R.dimen.column_header_height)

    private val rowHeight: Int = (height - headerColumnHeight) / ROWS_COUNT

    private val headerRowWidth: Int = context.resources.getDimensionPixelSize(R.dimen.row_header_width)

    private val columnWidth: Int = (width - headerRowWidth) / COLUMN_COUNT

    private val layoutInflater: LayoutInflater = LayoutInflater.from(context)

    private val rowsNames: Array<String> = context.resources.getStringArray(R.array.rows_names)

    //region create view holders

    override fun onCreateItemViewHolder(parent: ViewGroup): ViewHolderImpl =
        ScheduleItemViewHolder(layoutInflater.inflate(R.layout.item_schedule_cell, parent, false))

    override fun onCreateColumnHeaderViewHolder(parent: ViewGroup): ViewHolderImpl =
        ScheduleColumnHeaderViewHolder(layoutInflater.inflate(R.layout.item_schedule_column_header, parent, false) as TextView)

    override fun onCreateRowHeaderViewHolder(parent: ViewGroup): ViewHolderImpl =
        ScheduleRowHeaderViewHolder(layoutInflater.inflate(R.layout.item_schedule_row_header, parent, false) as TextView)

    override fun onCreateLeftTopHeaderViewHolder(parent: ViewGroup): ViewHolderImpl =
        ScheduleTopLeftViewHolder(layoutInflater.inflate(R.layout.item_schedule_top_left, parent, false))

    //endregion

    //region bind view holders

    override fun onBindLeftTopHeaderViewHolder(viewHolder: ViewHolderImpl) {
        //empty
    }


    override fun onBindHeaderColumnViewHolder(viewHolder: ViewHolderImpl, column: Int) {
        (viewHolder as ScheduleColumnHeaderViewHolder).setHeaderTitle(
            scheduleDataSource.getColumnHeaderDate(column).formatToTime()
        )
    }

    override fun onBindHeaderRowViewHolder(viewHolder: ViewHolderImpl, row: Int) {
        (viewHolder as ScheduleRowHeaderViewHolder).setHeaderTitle(
            String.format(rowsNames[row - 1], scheduleDataSource.getRowHeaderDate(row).formatToShortDate())
        )
    }

    override fun onBindViewHolder(viewHolder: ViewHolderImpl, row: Int, column: Int) {
        (viewHolder as ScheduleItemViewHolder).setChecked(scheduleDataSource.getScheduleCellItem(row, column)?.isSelected ?: false)
    }

    override fun getHeaderRowWidth(): Int = headerRowWidth

    override fun getRowHeight(row: Int): Int  = rowHeight

    override fun getColumnWidth(column: Int): Int = columnWidth

    override fun getHeaderColumnHeight(): Int = headerColumnHeight

    override fun getRowCount(): Int = ScheduleDataSource.ROWS_COUNT

    override fun getColumnCount(): Int = ScheduleDataSource.COLUMN_COUNT

}